codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parsers/JsonSerializerGenerator.kt

```diff
@@ -315,14 +330,16 @@ class JsonSerializerGenerator(protocolConfig: ProtocolConfig) : StructuredDataSe
     private fun RustWriter.serializeCollection(context: SimpleContext<CollectionShape>) {
         val itemName = safeName("item")
         rustBlock("for $itemName in ${context.valueExpression.asRef()}") {
-            serializeMember(
-                MemberContext(
-                    context.writerName,
-                    MemberDestination.Array(),
-                    ValueExpression.Reference(itemName),
-                    context.shape.member
-                )
+            val memberContext = MemberContext(
+                context.writerName,
+                MemberDestination.Array(),
+                ValueExpression.Reference(itemName),
+                context.shape.member
             )
+            when (context.shape.hasTrait<SparseTrait>()) {
```

Comment:
you could consider handling this more generally in `serializeMember`—the member type will actually be optional or not (you don't need to actually look at the sparse trait, just `memberSymbol.isOptional()`

I think that's what we do in XML

Reply:
Of course, you'd need to separately deal with the fact that only in collections is the `null` value actually serialized since it's ignored in structs.

Reply:
So maybe water bed complexity 🤷🏻

Reply:
But in general, it's better to let the symbol visitor centralize optionality. That allows us to, eg. have a config setting to turn sparse trait support on or off because the thing that we're actually generating on is the symbol